### PR TITLE
feat: integrate sentry

### PR DESCRIPTION
**Pull request by n-peugnet**
Maybe we should remove the Sentry package from the zip release as it's not really meant to be used by other people than you

This PR should be merged after #27.

### CODE REVIEW

**Comment by vincent-peugnet** [comment]
> Maybe we should remove the Sentry package from the zip release as it's not really meant to be used by other people than you

I don't really understand how it work, but if I understood what it's supposed to be, It sounds logical to me as well.

**Comment by n-peugnet** [comment]
Wait for this one it really isn't finished after all

File: Makefile [hunk 1]
```diff
@@ -3,80 +3,98 @@ include .env
...
-		composer.lock \
 		Makefile \
+		"composer*" \
 		"package*" \
 		webpack.config.js
-	zip -r $@ \
-		assets/js \
-		vendor \
-		-x "*test*" \
-		-x "*docs*"
```
Comment (n-peugnet) [inline Makefile]: The zip had permission issues when unzipping. I fixed that

File: Makefile [hunk 2]
```diff
@@ -3,80 +3,98 @@ include .env
...
 .env:
 	cp .default.env .env
 
+# use a force (fake) target to always rebuild this file but have Make
+# consider this updated if it was actually rewritten (a .PHONY target
+# is always considered new)
+VERSION: FORCE
+ifneq ($(CUR_VERSION),$(GIT_VERSION))
+	@echo $(GIT_VERSION) > VERSION
+endif
```
Comment (n-peugnet) [inline Makefile]: This makes the VERSION file

File: app/fn/fn.php [hunk 3]
```diff
@@ -76,7 +76,7 @@ function idclean(string $input)
 function getversion()
 {
 	if(file_exists('VERSION')) {
-		$version = file_get_contents('VERSION');
+		$version = trim(file_get_contents('VERSION'));
```
Comment (n-peugnet) [inline app/fn/fn.php]: now that VERSION is generated by make we need to trim it

File: app/view/templates/layout.php [hunk 4]
```diff
@@ -18,6 +18,13 @@
         echo '<link rel="stylesheet" href="' . Wcms\Model::csspath() . Wcms\Config::interfacecss() . '">';
     }
     ?>
+    <script>
+        const sentrydsn = '<?= Wcms\Config::sentrydsn() ?>';
+        const version = '<?= getversion() ?>';
+        const url = '<?= Wcms\Config::url() ?>';
+        const basepath = '<?= Wcms\Config::basepath() ?>';
+    </script>
+    <script src="<?= Wcms\Model::jspath() ?>sentry.bundle.js"></script>
```
Comment (n-peugnet) [inline app/view/templates/layout.php]: This is to enable Sentry in Javascript. It's added on every page and report javascript errors

File: composer.json [hunk 5]
```diff
@@ -6,7 +6,8 @@
         "altorouter/altorouter": "^1.2",
         "jamesmoss/flywheel": "^0.5.2",
         "league/plates": "^3.3",
-        "michelf/php-markdown": "^1.8"
+        "michelf/php-markdown": "^1.8",
+        "sentry/sdk": "^2.0"
```
Comment (n-peugnet) [inline composer.json]: new php dependency

File: index.php [hunk 6]
```diff
@@ -13,12 +13,24 @@
...
+    'dsn' => Wcms\Config::sentrydsn(),
+    'release' => getversion(),
+    'project_root' => 'app',
+]);
+Sentry\configureScope(function (Sentry\State\Scope $scope): void {
+    $scope->setUser([
+        'id' => Wcms\Config::url(),
+        'username' => Wcms\Config::basepath(),
+    ]);
+});
```
Comment (n-peugnet) [inline index.php]: This is to enable Sentry on PHP

File: index.php [hunk 7]
```diff
@@ -13,12 +13,24 @@
...
 
 try {
     $matchoper = new Wcms\Routes();
     $matchoper->match();
 
 } catch (Exception $e) {
+    Sentry\captureException($e);
```
Comment (n-peugnet) [inline index.php]: This is to report caught errors

File: package.json [hunk 8]
```diff
@@ -3,10 +3,11 @@
   "repository": "github:vincent-peugnet/wcms",
   "license": "MIT",
   "dependencies": {
+    "@sentry/browser": "^5.7.1",
     "codemirror": "^5.49.0"
   },
   "devDependencies": {
-    "@release-it/bumper": "^1.0.5",
+    "@sentry/cli": "^1.49.0",
```
Comment (n-peugnet) [inline package.json]: The CLI isused to create releases on Sentry. This is how we can upload the sourcemaps to Sentry

File: src/sentry.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import * as Sentry from '@sentry/browser';
+
+Sentry.init({
+    dsn: sentrydsn,
+    release: version,
+});
+Sentry.setUser({
+    id: url,
+    username: basepath,
+});
```
Comment (n-peugnet) [inline src/sentry.js]: The bundle that effectively starts Sentry on Js

File: webpack.config.js [hunk 10]
```diff
@@ -5,13 +5,14 @@ module.exports = (env) => {
 	return {
 		// Environment dependent
 		mode: env == 'dev' ? 'development' : 'production',
-		devtool: env == 'dev' ? 'inline-source-map' : 'none',
+		devtool: env == 'dev' ? 'cheap-eval-source-map' : 'source-map',
```
Comment (n-peugnet) [inline webpack.config.js]: Generate sourcemap files on prod build mode

**Comment by n-peugnet** [comment]
It still needs some documentation

**Comment by n-peugnet** [comment]
So here is a preview of the `make dist` resulting distribution zip.

It does not contain either the sourcemaps nor Sentry php packages.
The sentry js bundle is still present but it is not included in the pages.

[w_cms_v1.7.0-58-ge4bc654.zip](https://github.com/vincent-peugnet/wcms/files/3849042/w_cms_v1.7.0-58-ge4bc654.zip)

**Pull request merged**